kin-ecosystem-sdk/src/test/java/com/kin/ecosystem/data/blockchain/BlockchainSourceImplTest.java

```diff
@@ -0,0 +1,258 @@
+import org.mockito.InOrder;
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.mockito.MockitoAnnotations;
+import org.robolectric.RobolectricTestRunner;
+import org.robolectric.annotation.Config;
+
+@Config(manifest = Config.NONE)
+@RunWith(RobolectricTestRunner.class)
+public class BlockchainSourceImplTest {
+
+    private static final String PUBLIC_ADDRESS = "public_address";
+    private static final String APP_ID = "appID";
+    private static final String ORDER_ID = "orderID";
+    private static final String MEMO_EXAMPLE = "1-" + APP_ID + "-" + ORDER_ID;
+
+
+    @Mock
+    private KinClient kinClient;
+
+    @Mock
+    private BlockchainSource.Local local;
+
+    @Mock
+    private KinAccount kinAccount;
+
+    @Mock
+    private BlockchainEvents blockchainEvents;
+
+
+    @Captor
+    private ArgumentCaptor<EventListener<Void>> accountCreationCaptor;
+
+    @Mock
+    private ListenerRegistration accountRegistration;
+
+
+    @Mock
+    private Request<Void> activateAccountReq;
+
+    @Captor
+    private ArgumentCaptor<ResultCallback<Void>> accountActivateCaptor;
+
+
+    @Mock
+    private Request<Balance> getBalanceReq;
+
+    @Captor
+    private ArgumentCaptor<ResultCallback<Balance>> getBalanceCaptor;
+
+    @Mock
+    private Balance balanceObj;
+
+
+    private BlockchainSourceImpl blockchainSource;
+    private BalanceUpdate balanceUpdate;
+    private int x = 0;
+
+    @Before
+    public void setUp() throws Exception {
+        MockitoAnnotations.initMocks(this);
+        when(kinClient.addAccount()).thenReturn(kinAccount);
+        when(kinAccount.blockchainEvents()).thenReturn(blockchainEvents);
+        when(kinAccount.getBalance()).thenReturn(getBalanceReq);
+        when(kinAccount.activate()).thenReturn(activateAccountReq);
+        when(blockchainEvents.addAccountCreationListener(any(EventListener.class))).thenReturn(accountRegistration);
+        when(balanceObj.value()).thenReturn(new BigDecimal(20));
+        when(kinAccount.getPublicAddress()).thenReturn(PUBLIC_ADDRESS);
+
+        resetInstance();
+        resetPaymentObserverCount();
+        resetBalanceObserverCount();
+
+        // Account Creation
+        verify(kinClient).addAccount();
+        verify(blockchainEvents).addAccountCreationListener(accountCreationCaptor.capture());
+        accountCreationCaptor.getValue().onEvent(null);
+
+        verify(activateAccountReq).run(accountActivateCaptor.capture());
+        accountActivateCaptor.getValue().onResult(null);
+
+        // init Balance
+        verify(getBalanceReq).run(getBalanceCaptor.capture());
+        getBalanceCaptor.getValue().onResult(balanceObj);
+        verify(local).setBalance(balanceObj.value().intValue());
+
+        when(kinClient.getAccount(0)).thenReturn(kinAccount);
+        balanceUpdate = new BalanceUpdate();
+    }
+
+
+    private void resetInstance() throws Exception {
+        Field instance = BlockchainSourceImpl.class.getDeclaredField("instance");
```

Comment:
I think it's better to make ctor internal instead of private, this way you can instantiate it freely inside tests instead of dealing with static reference, the danger with exposing it inside the package seems small and negligible

Reply:
not sure i understand, lets talk